sshd-core/src/main/java/org/apache/sshd/client/auth/pubkey/UserAuthPublicKey.java

```diff
@@ -56,6 +58,7 @@
     protected Iterator<PublicKeyIdentity> keys;
     protected PublicKeyIdentity current;
     protected List<NamedFactory<Signature>> factories;
+    protected List<String> currentAlgorithms = new LinkedList<>();
```

Comment:
A `LinkedList` is not really a `List` and any attempt to use it as such for random access has horrible performance (true for large lists but let's not quibble). Since the code only handles the "edges" (head/tail) I re commend defining it as
```java
Deque<String> currentAlgorithms = new LinkedList<>();
```
and the you can use `removeHead/addTail`/etc. And since `Deque extends Collection` you can also use it as such.

Reply:
Done.

Reply:
Great - what about  the comparison between the requested signature and the one returned by the server + the warning log of their mismatch ?

Reply:
Fair enough - in that case I have no further comments